DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/ViewTreeRecordingContext.swift

```diff
@@ -0,0 +1,89 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+import Foundation
+import SafariServices
+
+/// The context of recording subtree hierarchy.
+///
+/// Some fields are mutable, so `NodeRecorders` can specialise it for their subtree traversal.
+internal struct ViewTreeRecordingContext {
+    /// The context of the Recorder.
+    let recorder: Recorder.Context
+    /// The coordinate space to convert node positions to.
+    let coordinateSpace: UICoordinateSpace
+    /// Generates stable IDs for traversed views.
+    let ids: NodeIDGenerator
+    /// Provides base64 image data with a built in caching mechanism.
+    let imageDataProvider: ImageDataProviding
+    /// Available text obfuscators to use accordingly to current privacy mode.
+    let textObfuscators: TextObfuscators
+    /// Variable view controller related context
+    var viewControllerContext: ViewControllerContext = .init()
+}
+
+internal extension ViewTreeRecordingContext {
+    /// The `ViewControllerContext` struct is used for storing context-related information about the parent view controller and its type.
+    struct ViewControllerContext {
+        /// An enumeration representing the different types of view controllers that we handle in special way.
+        enum ViewControllerType {
+            case alert
+            case safari
+            case activity
+            case other
+
+            /// An initializer that takes a `UIViewController` and determines its corresponding `ViewControllerType`.
+            ///
+            /// - Parameter viewController: The `UIViewController` for which to determine the `ViewControllerType`.
+            internal init(_ viewController: UIViewController?) {
+                switch viewController {
+                case is UIAlertController:
+                    self = .alert
+                case is UIActivityViewController:
+                    self = .activity
+                case is SFSafariViewController:
+                    self = .safari
+                default:
+                    self = .other
+                }
+            }
+        }
+
+        /// The parent view controller's type.
+        var parentType: ViewControllerType?
+
+        /// A boolean flag indicating whether the current view is the root view or not.
+        var isRootView: Bool?
```

Comment:
question/ What would `nil` value mean for this boolean flag? Looks that it should be non-optional.